### PR TITLE
Expose and test (I)Dictionary.GetValueOrDefault

### LINKED ISSUE #3482: Proposal: (I)Dictionary<TK,TV>.GetValueOrDefault
## Scenario

In almost every app or library I write I almost always end up needing to write code like this, in at least one place:

``` c#
SomeClass returnValue;
return DictionaryOfSomeClass.TryGetValue(someKey, out returnValue) ? returnValue : null;
```

The reason that the `IDictionary<TKey, TValue>` contract throws from its indexer for a non-existent key are well-known: an alternative behavior of returning `null` is nonsensical if `TValue` is a value type. Returning `default(TValue)` would work, but would clearly be a misleading behavior.

But with dictionaries of class-types it is often desirable and expected to return `null` for a non-existent key.

## Proposal

```c#
namespace System.Collections.Generic
{
    public class Dictionary<TKey, TValue>
    {
        public TValue GetValueOrDefault(TKey key);
        public TValue GetValueOrDefault(TKey key, TValue defaultValue);
    }

    public static class CollectionExtensions
    {
        public static TValue GetValueOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key);
        public static TValue GetValueOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key, TValue defaultValue);

        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key);
        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key, TValue defaultValue);
    }
}
```

**Notes:**
- There will be source-compile problem with ambiguity in all projects that already have such custom extension method `GetValueOrDefault` on `IDictionary` or `IReadOnlyDictionary`. Which is general problem for any extension method we add into CoreFX.

### Original Proposal - replaced by the one above

I propose the following extension method (I don't like the name `GetValueNoThrow`, so that can be ignored for now). Please note that it applies only when `TValue` is a class, so it does not violate any behavior for value types.

``` c#
    public static class ClassyDictionaryExtensions
    {
        public static TValue GetValueNoThrow_Ignore_The_Name<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key)
            where TValue : class
        {
            if (dictionary == null)
            {
                throw new ArgumentNullException(nameof(dictionary));
            }
            TValue value;
            dictionary.TryGetValue(key, out value);
            return value;
        }
    }
```

The new code would end up with one very-easy-to-read line of code:

``` c#
            return DictionaryOfSomeClass.GetValueNoThrow(someKey);
```
## Discussion

Would others find such an extension method useful?

### PR DESCRIPTION
**Pull request by hughbe**
Running `msbuild src\System.Collections\src\System.Collections.csproj /p:TargetGroup=uapaot /p:BaselineAllApiCompatError=true` doesn't seem to change anything but I hope this passes CI

Fixes #3482

### CODE REVIEW

File: src/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                throw new ArgumentNullException(nameof(dictionary));
+            }
+
+            TValue value;
+            if(dictionary.TryGetValue(key, out value))
+            {
+                return value;
+            }
+
+            return defaultValue;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs]: Nit: I'd personally have written this as
```C#
return dictionary.TryGetValue(key, out value) ? value : defaultValue;
```
but what you have is fine.

File: src/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs]: Same nits

**Comment by hughbe** [comment]
Addressed all your feedback, thanks!

File: src/System.Collections/src/ApiCompatBaseline.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Compat issues with assembly System.Collections:
```
Comment (jkotas) [inline src/System.Collections/src/ApiCompatBaseline.txt]: This should not be a universal ApiCompatBaseline.txt. It should be specific to just uap.

**Comment by karelz** [comment]
Thank you for your fix @hughbe! Really appreciate your help.
Please ping the issues when you grab them, so that we can assign them to you ...

File: src/System.Collections/tests/Generic/CollectionExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        [Fact]
+        public void GetValueOrDefault_NullIReadOnlyDictionary_ThrowsArgumentNullException()
+        {
+            IReadOnlyDictionary<string, string> dictionary = null;
+            Assert.Throws<ArgumentNullException>("dictionary", () => dictionary.GetValueOrDefault("key"));
+            Assert.Throws<ArgumentNullException>("dictionary", () => dictionary.GetValueOrDefault("key", "value"));
+        }
+    }
+}
```
Comment (hughbe) [inline src/System.Collections/tests/Generic/CollectionExtensionsTests.cs]: Fixed, used SortedSet

**Comment by chester89** [comment]
@karelz did this method make it to 4.7.2? Looks like it should, but MSDN says otherwise

**Comment by karelz** [comment]
It was actually reverted back - see PR #17917.
There are extension methods instead in [`CollectionExtensions`](https://apisof.net/catalog/System.Collections.Generic.CollectionExtensions.GetValueOrDefault%3CTKey,TValue%3E(IReadOnlyDictionary%3CTKey,TValue%3E,TKey)). They didn't make it into .NET Framework. We do not have plans to add them there.

**Comment by chester89** [comment]
I see. Thanks for help

**Pull request merged**